src/modules/mavlink/streams/GPS_GLOBAL_ORIGIN.hpp

```diff
@@ -0,0 +1,103 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <uORB/topics/vehicle_local_position.h>
+
+class MavlinkStreamGpsGlobalOrigin : public MavlinkStream
+{
+public:
+	const char *get_name() const override
+	{
+		return MavlinkStreamGpsGlobalOrigin::get_name_static();
+	}
+
+	static constexpr const char *get_name_static()
+	{
+		return "GPS_GLOBAL_ORIGIN";
+	}
+
+	static constexpr uint16_t get_id_static()
+	{
+		return MAVLINK_MSG_ID_GPS_GLOBAL_ORIGIN;
+	}
+
+	uint16_t get_id() override
+	{
+		return get_id_static();
+	}
+
+	static MavlinkStream *new_instance(Mavlink *mavlink)
+	{
+		return new MavlinkStreamGpsGlobalOrigin(mavlink);
+	}
+
+	unsigned get_size() override
+	{
+		return _vehicle_local_position_sub.advertised() ?
+		       (MAVLINK_MSG_ID_GPS_GLOBAL_ORIGIN_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES) : 0;
+	}
+
+private:
+	uORB::Subscription _vehicle_local_position_sub{ORB_ID(vehicle_local_position)};
+
+	/* do not allow top copying this class */
+	MavlinkStreamGpsGlobalOrigin(MavlinkStreamGpsGlobalOrigin &) = delete;
+	MavlinkStreamGpsGlobalOrigin &operator = (const MavlinkStreamGpsGlobalOrigin &) = delete;
+
+protected:
+	explicit MavlinkStreamGpsGlobalOrigin(Mavlink *mavlink) : MavlinkStream(mavlink)
+	{}
+
+	bool send() override
+	{
+		vehicle_local_position_s vehicle_local_position;
+
+		if (_vehicle_local_position_sub.update(&vehicle_local_position)) {
+			mavlink_gps_global_origin_t msg{};
+
+			if (vehicle_local_position.ref_timestamp > 0) {
+				msg.latitude = static_cast<int32_t>(vehicle_local_position.ref_lat * 1e7); // double degree -> int32 degreeE7
+				msg.longitude = static_cast<int32_t>(vehicle_local_position.ref_lon * 1e7); // double degree -> int32 degreeE7
+				msg.altitude = static_cast<int32_t>(vehicle_local_position.ref_alt * 1e3f); // float m -> int32 mm
+				msg.time_usec = vehicle_local_position.timestamp; // int64 time since system boot
+			}
+
+			mavlink_msg_gps_global_origin_send_struct(_mavlink->get_channel(), &msg);
```

Comment:
Doesn't this need to go inside the conditional or are you intentionally sending the zeroed message when there's no valid origin?

Reply:
I'm intentionally sending it. I tried to simplify now in https://github.com/PX4/PX4-Autopilot/pull/16204/commits/9dc4231142cfa12d5107b53546e813157ae47686 according to your last review.

Reply:
I don't think it should be sent before there's a global origin unless the Mavlink message mentions how to handle invalidating the fields. 

http://mavlink.io/en/messages/common.html#GPS_GLOBAL_ORIGIN

Reply:
I was thinking what if I request the origin and I get nothing back? Also is the use case fixed that once a reference is set there's always one?

Reply:
ok, I changed it. How about now? https://github.com/PX4/PX4-Autopilot/pull/16204/commits/37921a2e13812453d53fa4acb8149d4ae6569561

Reply:
Oh hold on, we might be thinking about this differently.

I was assuming that MAVSDK (or anyone) would be requesting to start the `GPS_GLOBAL_ORIGIN` stream at some low rate (`MAV_CMD_SET_MESSAGE_INTERVAL`). 

The other option is `MAV_CMD_REQUEST_MESSAGE` which is kind of the same thing, but only sends the message once.


> I was thinking what if I request the origin and I get nothing back?

Adding some mechanism in `GPS_GLOBAL_ORIGIN` to indicate there's no valid origin would probably be the right thing to do, but I know getting changes into Mavlink is time consuming.

In Mavlink I don't see any actual requirement that you must respond to `MAV_CMD_REQUEST_MESSAGE`. 

> Also is the use case fixed that once a reference is set there's always one?

Right now effectively yes, but I'm not sure if in the future there could be valid cases where it's lost.

Reply:
> ok, I changed it. How about now? [37921a2](https://github.com/PX4/PX4-Autopilot/commit/37921a2e13812453d53fa4acb8149d4ae6569561)

To me this seems like the right thing to do, but maybe @julianoes or @JonasVautherin have a different opinion from the Mavlink and MAVSDK perspective?